Fixes #22156: Fix Spark3 runner to compile against Spark 3.2/3.3

mosche opened the pull request:
Summary of changes:
- Add tests to verify compatibility with Spark 3.2 and 3.3 (3.1.2 remains the default).
- Use correct group for Dropwizard `metrics-core`. While package names remained the same, modules have moved to a new group `io.dropwizard.metrics`. This group is also used by Spark and leads to rather indeterministic behavior when loading classes as outdated classes of the old group remain on the classpath.
 -  Recent versions of Dropwizard `MetricRegistry` do not allow registering arbitrary implementations of `Metric`s. Migrate to using `Gauge`.
- Spark `Sink`s have different constructors between 3.1 and later versions. Using reflection here is fine as there's just one sink per executor.
- Change encoders/expressions for StructuredStreamingRunner to use reflection where necessary. The overhead of this is rather low as it's just necessary once to build the execution plan.
- Change StructuredStreamingRunner to allow using a provided / active Spark Sessions for tests. This was critical to fix the broken tests for the metrics integration. On shutdown Spark will stop the metrics system, after that point it's not possible to query any metrics (from the in-memory sink) anymore.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

mosche commented:
R: @aromanenko-dev 
R: @echauchot

echauchot commented:
reviewing ...

mosche commented:
@echauchot any comments / feedback already? we can also take this in turns, so you don't have to review everything at once

echauchot commented:
> @echauchot any comments / feedback already? we can also take this in turns, so you don't have to review everything at once
@mosche I was off on thrusday and friday so I have a dozen files left to review (only the metrics changes). I'll submit all the comments as a batch when I'm done with the review

echauchot reviewed (commented):
@mosche thanks for your work. Here is a partial review of all except the metrics changes.

mosche commented on runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/helpers/EncoderFactory.java:
```diff
@@ -17,33 +17,48 @@
+   * Invoke method {@code fun} on Class {@code cls}, immediately propagating {@code null} if any
+   * input arg is {@code null}.
+   *
+   * <p>To address breaking interfaces between various version of Spark 3 these are created
```
I tried doing that initially and decided it's doing more harm than good it as it required passing the spark session allround, including static encoder factories and stuff like that.  I'll keep as is.

echauchot replied:
I meant extract the spark version at the beginning of the pipeline and pass only an int to the methods.

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/SparkStructuredStreamingRunner.java:
```diff
@@ -146,10 +146,12 @@ public SparkStructuredStreamingPipelineResult run(final Pipeline pipeline) {
+    Runnable onTerminalState =
+        options.getUseActiveSparkSession()
+            ? () -> {}
+            : () -> translationContext.getSparkSession().stop();
```
That transition doesn't exist anymore, the only way to get here is through `offerNewState` which by the time of this check already replaced `state` with `newState` (which is a terminal state and never `RUNNING`).
This was previously only used to expose `stop()` for testing purposes.

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/sink/CsvSink.java:
```diff
@@ -18,22 +18,64 @@
+
+  private static org.apache.spark.metrics.sink.CsvSink newDelegate(Object... params) {
+    try {
+      Constructor<?> constructor = org.apache.spark.metrics.sink.CsvSink.class.getConstructors()[0];
```
👍 I can change that if you prefer, but it doesn't provide any additional safety / checks. That's the version i used initially. These constructors are only ever called by Spark's `MetricsSystem` (using reflection btw), so having anything else other than the default constructor would be just pointless / dead code.
Also, in both cases `newInstance` uses varargs and it will always fail at runtime without any compilation error.

mosche commented on runners/spark/src/test/java/org/apache/beam/runners/spark/structuredstreaming/SparkSessionRule.java:
```diff
@@ -37,23 +39,43 @@ public SparkSessionRule(String sparkMaster, Map<String, String> sparkConfig) {
     this(sparkMaster, Arrays.stream(sparkConfig).collect(toMap(KV::getKey, KV::getValue)));
   }
 
+  private static int numDriverCores(String master) {
```
This has become obsolete with above comment, introduced a `SparkSessionFactory`. Btw, that's the Spark terminology... see `SparkContext.numDriverCores(String master, ...)`

echauchot replied:
+1 on the factory, I just looked at it

mosche commented on runners/spark/src/test/java/org/apache/beam/runners/spark/structuredstreaming/SparkSessionRule.java:
```diff
@@ -37,23 +39,43 @@ public SparkSessionRule(String sparkMaster, Map<String, String> sparkConfig) {
 
   public SparkSessionRule(KV<String, String>... sparkConfig) {
-    this("local", sparkConfig);
+    this("local[2]", sparkConfig);
```
If you prefer using `local[4]` I'm fine changing. That's just a value taken from past experience having worked lots on improving performance of Spark tests. 
For tests it's generally a misconception that more cores (e.g. all available cores) mean more performance. Tests are typically running in parallel in multiple forks. More cores also means more congestion in that case. Also, using more partitions means more overhead on the processing side (always assuming tests operate on rather tiny datasets). That's why it's important to reduce the number of shuffle partitions for tests before Spark 3.2. (looks like later Spark versions are smarter about this).
On the other side, using just 1 core / 1 partition can be dangerous as it might obfuscate problems (e.g. invalid assumptions about ordering) in the code.

mosche commented on runners/spark/2/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/helpers/EncoderFactory.java:
```diff
@@ -17,38 +17,35 @@
-        classTag);
+        serializers,
+        deserializer,
+        ClassTag$.MODULE$.apply(clazz));
```
Unfortunately not, that's a breaking change in the Scala version

echauchot replied:
arf, no luck

mosche commented on runners/spark/src/test/java/org/apache/beam/runners/spark/structuredstreaming/translation/helpers/EncoderHelpersTest.java:
```diff
@@ -35,13 +45,52 @@ public class EncoderHelpersTest {
   }
+
+  @Test
+  public void testBeamEncoderOfPrivateType() {
```
private types cannot be used in code generation, when trying to compile such code it would throw an `IllegalAccessError`.

echauchot replied:
ok

mosche commented on runners/spark/2/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/helpers/EncoderFactory.java:
```diff
@@ -17,38 +17,35 @@
-
+  static <T> Encoder<T> create(
+      Expression serializer, Expression deserializer, Class<? super T> clazz) {
+    List<Expression> serializers = Nil$.MODULE$.$colon$colon(serializer);
```
I agree it's ugly and painful to interface with Scala from Java. BUT that's also the nature of dealing with Spark at a lower level beyond what's offered on the Java API to create an efficient runner :(
What I'd suggest is to create a single static utility `ScalaInterop(erability)` to deal with these. But I'd rather not do it in this PR as I'm running into more and more conflicts with the pending PR for the structured streaming runner.

echauchot replied:
I agree it is better to leave as it is and open a ticket to improve scala interrop later on. Please open this ticket, assign it to you and refer it in this PR.

mosche replied:
👍

echauchot replied:
have you opened the ticket already? If so, please refer it here

mosche replied:
https://github.com/apache/beam/issues/22382

mosche commented:
Thanks a lot for the first round of feedback @echauchot, I've pushed related changes.

mosche commented:
Hmm, looks like license headers for shared sources aren't added automatically ...

echauchot reviewed (changes requested):
@mosche I finished the review of all the files. I'll now take a look at the comments you addressed from round 1.

echauchot commented:
Run Spark StructuredStreaming ValidatesRunner

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/BeamMetricSet.java:
```diff
@@ -0,0 +1,56 @@
+    return new StaticGauge(number.doubleValue());
+  }
+
+  private static class StaticGauge implements Gauge<Double> {
```
Intention of that name was a different one here... a Gauge is typical something that pulls a value somewhere, this one is rather static (or better constant) in behavior.  How about one of these: `StaticValue`, `ConstantGauge`, `ConstGauge`, `ConstantValue`, `ConstValue`, `Constant`?

echauchot replied:
ah I misunderstood the intent of the name. ConstantGauge lokks fine to me

mosche replied:
👍

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/BeamMetricSet.java:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  private static class StaticGauge implements Gauge<Double> {
+    double value;
```
@mosche make final for clarity

mosche replied:
done

echauchot replied:
it is indeed even clearer with the final that we deal with a constant gauge

echauchot commented on .test-infra/jenkins/job_PreCommit_Java_Spark3_Versions.groovy:
```diff
@@ -0,0 +1,37 @@
+    nameBase: 'Java_Spark3_Versions',
+    gradleTask: ':runners:spark:3:sparkVersionsTest',
+    gradleSwitches: [
+      '-PdisableSpotlessCheck=true'
```
agree, fair enough

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/AggregatorMetric.java:
```diff
@@ -33,7 +41,35 @@ public static AggregatorMetric of(final NamedAggregators namedAggregators) {
+    return metrics;
+  }
+
+  // Metric type is assumed to be compatible with Double
```
Yes 👍 Among the know supported subtypes of `Metric` only `Gauge` and `MetricRegistry` fit. `MetricRegistry` would be even better, but also require a lot more changes. Spark 3 has a plugin framework which would simplify a lot of this (no custom sinks needed anymore + configuration nightmare for users). When migrating to a metrics plugin, this needs to turn into a registry. One thing at a time ;)

echauchot replied:
please open a ticket for later on this matter.

mosche replied:
👍  https://github.com/apache/beam/issues/22384

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkBeamMetric.java:
```diff
@@ -33,61 +37,71 @@
+  public Map<String, Gauge<Double>> getValue(String prefix, MetricFilter filter) {
+    MetricResults metricResults =
+        asAttemptedOnlyMetricResults(MetricsAccumulator.getInstance().value());
+    Map<String, Gauge<Double>> metrics = new HashMap<>();
```
If you have a look at `WithMetricsSupport` you can see that everything was previously returned as `Gauge` as well. Nothing has changed in that regard, this part was just a bit of a code cleanup on the way. For distributions the individual values are extracted as gauge each exactly the same was as before.

echauchot replied:
yes, I reviewed the `WithMetricsSupport` after I wrote this comment and I forgot to update it.

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkBeamMetric.java:
```diff
@@ -33,61 +37,71 @@
-
-    ArrayList<String> pieces = new ArrayList<>();
+    return Streams.concat(
+            Stream.of(prefix), // prefix is not cleaned, should it be?
```
it's the same as before to avoid any breaking change ...

echauchot replied:
in the default case prefix == "" ?

mosche replied:
yes and no, usage is always as follows (`prefix = entry.getKey()`)
```
((BeamMetricSet) gauge).getValue(entry.getKey(), filter)
```

echauchot replied:
:+1:

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkBeamMetric.java:
```diff
@@ -33,61 +37,71 @@
-    ArrayList<String> pieces = new ArrayList<>();
+    return Streams.concat(
+            Stream.of(prefix), // prefix is not cleaned, should it be?
+            Stream.of(stripSuffix(cleanPart(key.stepName()))),
```
but what's the value of keeping the step variable?

echauchot commented on runners/spark/3/build.gradle:
```diff
@@ -28,3 +28,37 @@ project.ext {
+    "321": "3.2.1"
+]
+
+sparkVersions.each { kv ->
```
So, the runner is built against 3.1.2 but can run on 3.2 and 3.3 ? This is why you need to dynamically get the spark version either by reflexion or by getting it from the spark session. Right ?

mosche replied:
exactly... that's pretty much how things look like in real life. you run a job build with one version of beam on a spark cluster, but the spark versions the runner was compiled against and the cluster don't necessarily match

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkBeamMetric.java:
```diff
@@ -33,61 +37,71 @@
+  }
 
-    return String.join(".", pieces);
+  private void putFiltered(
```
filters got applied previously, but at the very end after rendering everything. pushing these down just avoids unnecessary work, see https://github.com/apache/beam/pull/22157/files/388512454d4ede037a7b67f2d33d0198a494cff4#diff-91b7484788c580d5c170c9c1add3a7f32fa0a7e2eeb00a6fd4e93db0508bb9c0L119

echauchot replied:
yes, Once again I saw the `WithMetricsSupport` class after this comment...

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/metrics/sink/CodahaleCsvSink.java:
```diff
@@ -21,16 +21,66 @@
+ * }</pre>
  */
-public class CodahaleCsvSink extends org.apache.spark.metrics.sink.CsvSink {
+public class CodahaleCsvSink implements Sink {
```
I don't wanna introduce any breaking changes in this PR

echauchot replied:
let's keep this way to avoid silent breaking change

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/metrics/sink/CodahaleGraphiteSink.java:
```diff
@@ -21,14 +21,69 @@
+ * "spark.metrics.conf.*.sink.graphite.regex"="<optional_regex_to_send_matching_metrics>"
+ * }</pre>
+ */
+public class CodahaleGraphiteSink implements Sink {
```
I don't wanna introduce any breaking changes in this PR

echauchot replied:
need to think. It is not very user facing except for configuration like "my sink = my class"

mosche replied:
well, still a breaking change (probably even worse as this will cause it to fail silently at runtime)

echauchot replied:
silently, with not even a ClassNotFoundException ?

mosche replied:
That exception is caught by Spark and logged as error, but very easy to miss

echauchot replied:
ok so let's keep this way and avoid breaking change

echauchot commented on runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/helpers/EncoderHelpers.java:
```diff
@@ -19,256 +19,53 @@
+    Class<? super T> clazz = coder.getEncodedTypeDescriptor().getRawType();
+    // Class T could be private, therefore use OBJECT_TYPE to not risk an IllegalAccessError
+    return EncoderFactory.create(
+        beamSerializer(rootRef(OBJECT_TYPE, true), coder),
+        beamDeserializer(rootCol(BinaryType), coder),
```
and no perf drop noticed to use the higher level API ?

mosche replied:
absolutely not, it compiles to the same code (just invoking the static coder utils)

echauchot replied:
:+1:

mosche commented on runners/spark/src/test/java/org/apache/beam/runners/spark/structuredstreaming/aggregators/metrics/sink/SparkMetricsSinkTest.java:
```diff
@@ -21,51 +21,42 @@
- */
-@Ignore("Has been failing since at least c350188ef7a8704c7336f3c20a1ab2144abbcd4a")
+/** A test that verifies Beam metrics are reported to Spark's metrics sink in batch mode. */
 @RunWith(JUnit4.class)
```
what can be removed?

echauchot replied:
no the RunWith annotation I think

mosche replied:
jepp, i removed it

echauchot replied:
it is removed from everywhere except SparkMetricsSinkTest in the SS runner

mosche replied:
ok, thx, fixed that 👍

echauchot commented:
@mosche as a general advice, it is easier for review process if you do tiny commits that address one comment (we will squash anyway) that way I can track the addressed comments and resolve the conversations myself

mosche commented:
> as a general advice, it is easier for review process if you do tiny commits that address one comment (we will squash anyway) that way I can track the addressed comments and resolve the conversations myself

fair comment @echauchot, but I guess we have to meet somewhere in the middle ... that would make addressing comments more than a hassle

echauchot commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/BeamMetricSet.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Map;
+
+/**
+ * Map of Beam metrics available from {@link Gauge#getValue()}.
+ *
+ * <p>Note: Recent versions of Dropwizard {@link com.codahale.metrics.MetricRegistry MetricRegistry}
+ * do not allow registering arbitrary implementations of {@link com.codahale.metrics.Metric
+ * Metrics}.
```
ping ?

echauchot commented:
> > as a general advice, it is easier for review process if you do tiny commits that address one comment (we will squash anyway) that way I can track the addressed comments and resolve the conversations myself
> 
> fair comment @echauchot, but I guess we have to meet somewhere in the middle ... that would make addressing comments more than a hassle

yes somewere in the middle looks good to me. What I advice is what I do when I'm being reviewed but it is true that sometimes I commit more than one change and it requires that I split the commit afterwards. Painful ...

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/BeamMetricSet.java:
```diff
@@ -0,0 +1,56 @@
+ * do not allow registering arbitrary implementations of {@link com.codahale.metrics.Metric
+ * Metrics}.
+ */
+public abstract class BeamMetricSet implements Gauge<Map<String, Gauge<Double>>> {
```
👍

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkBeamMetric.java:
```diff
@@ -33,61 +37,71 @@
-      }
-      pieces.add(step);
-    }
+  private static @Nullable String cleanPart(@Nullable String str) {
```
👍

mosche commented on runners/spark/src/test/java/org/apache/beam/runners/spark/structuredstreaming/aggregators/metrics/sink/SparkMetricsSinkTest.java:
```diff
@@ -21,51 +21,42 @@
   @Rule public ExternalResource inMemoryMetricsSink = new InMemoryMetricsSinkRule();
 
+  @Rule
+  public TestPipeline pipeline = TestPipeline.fromOptions(sessionRule.createPipelineOptions());
```
👍

mosche commented on runners/spark/src/test/java/org/apache/beam/runners/spark/structuredstreaming/aggregators/metrics/sink/SparkMetricsSinkTest.java:
```diff
@@ -76,9 +67,10 @@ public void testInBatchMode() throws Exception {
+
     PAssert.that(output).containsInAnyOrder(EXPECTED_COUNTS);
-    pipeline.run();
+    pipeline.run().waitUntilFinish();
```
👍

echauchot replied:
`waitUntilFish` is still the default with `testMode` on you mean ? So if testMode is activated why need to explicitely call `waitUntilFinish()` here ?

mosche replied:
it's not need, already removed

echauchot replied:
:+1:

mosche commented:
@echauchot I've pushed all changes, pls have another look.

echauchot commented on runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/AggregatorMetric.java:
```diff
@@ -33,7 +41,35 @@ public static AggregatorMetric of(final NamedAggregators namedAggregators) {
+      String name = prefix + "." + entry.getKey();
+      Object rawValue = entry.getValue();
+      if (rawValue == null) {
+        continue;
```
ping ?

mosche replied:
pushed already

mosche replied:
also, that's pretty much personal taste, let's skip such stuff in the future

echauchot replied:
yes agree. And moreover I see this continue pattern all over the Beam code base. So it is definitely accepted by the community. I'm not a fan of it, but I will no more mention it

echauchot commented on runners/spark/src/test/java/org/apache/beam/runners/spark/metrics/SparkBeamMetricTest.java:
```diff
@@ -27,6 +27,7 @@
 
 /** Test SparkBeamMetric. */
 public class SparkBeamMetricTest {
```
ping ?

mosche replied:
done and pushed 👍

mosche commented on runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/metrics/BeamMetricSet.java:
```diff
@@ -0,0 +1,56 @@
+ * do not allow registering arbitrary implementations of {@link com.codahale.metrics.Metric
+ * Metrics}.
+ */
+public abstract class BeamMetricSet implements Gauge<Map<String, Gauge<Double>>> {
```
👍

echauchot commented:
Run Spark ValidatesRunner

echauchot reviewed (approved):
Thanks for your work Mortitz ! LGTM modulo passing tests. Ping me when everything passes so that I merge the PR

mosche commented:
Run Java PreCommit

mosche commented:
@echauchot tests are green
